### ISSUE TITLE
Printing ranges displays different behavior using writeln and writef

### ISSUE THREAD

**Issue by jabraham17** [chapel-lang/chapel#22250]
### Summary of Problem

When printing a range with alignment or stride that border the min/max of a integer width, `writeln` and `writef` behave differently and print the range with different values on M1 MacBooks. The issue does not occur on Intel MacBooks or regular x86 machines.

This problem was made more prevalent with #22225, but exists in prior versions of Chapel

### Steps to Reproduce

excerpt from `test/types/range/minMaxStrides.chpl`

Although the first writef should be equivalent to the send writef+writeln, they are different.

```
config var percentT=true;
proc test(type t) {
  type ST = int(Types.numBits(t));
  type UT = uint(Types.numBits(t));
  writeln("Testing signed ", ST:string, " and unsigned ", UT:string);

  for by_ in (Types.min(ST), Types.max(ST), Types.max(ST):UT) {
    for align_ in (Types.min(ST), Types.max(ST), Types.max(UT)) {
      var r = (0:align_.type)..(100:align_.type) by by_ align align_;
      if percentT then 
        writef("by_=%12n:%-10s; align_=%12n:%-10s; range=%t\n",  by_, by_.type:string, align_, align_.type:string, r);
      else { 
        writef("by_=%12n:%-10s; align_=%12n:%-10s; range=",  by_, by_.type:string, align_, align_.type:string);
        writeln(r);
      }
    }
  }
}

test(int(8));
```

**Compile command:**
`chpl test.chpl`

**Execution command:**
`./test` produces 
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; range=0..100 by -128 align 0
by_=        -128:int(8)    ; align_=         127:int(8)    ; range=0..100 by -128 align 0
by_=        -128:int(8)    ; align_=         255:uint(8)   ; range=0..100 by -128 align 1
by_=         127:int(8)    ; align_=        -128:int(8)    ; range=0..100 by 127 align 1
by_=         127:int(8)    ; align_=         127:int(8)    ; range=0..100 by 127
by_=         127:int(8)    ; align_=         255:uint(8)   ; range=0..100 by 127 align 1
by_=         127:uint(8)   ; align_=        -128:int(8)    ; range=0..100 by 127 align 1
by_=         127:uint(8)   ; align_=         127:int(8)    ; range=0..100 by 127
by_=         127:uint(8)   ; align_=         255:uint(8)   ; range=0..100 by 127 align 1
```

`./test --percentT=false` produces
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; range=0..100 by -128 align 0
by_=        -128:int(8)    ; align_=         127:int(8)    ; range=0..100 by -128 align 127
by_=        -128:int(8)    ; align_=         255:uint(8)   ; range=0..100 by -128 align 127
by_=         127:int(8)    ; align_=        -128:int(8)    ; range=0..100 by 127 align 126
by_=         127:int(8)    ; align_=         127:int(8)    ; range=0..100 by 127
by_=         127:int(8)    ; align_=         255:uint(8)   ; range=0..100 by 127 align 1
by_=         127:uint(8)   ; align_=        -128:int(8)    ; range=0..100 by 127 align 126
by_=         127:uint(8)   ; align_=         127:int(8)    ; range=0..100 by 127
by_=         127:uint(8)   ; align_=         255:uint(8)   ; range=0..100 by 127 align 1
```

#### In Chapel 1.30.0

`./test` produces 
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; range=0..100 by -128 align 1
by_=        -128:int(8)    ; align_=         127:int(8)    ; range=0..100 by -128 align 1
by_=        -128:int(8)    ; align_=         255:uint(8)   ; range=0..100 by -128(null):64: error: halt reached - casting int(8) less than 0 to uint(8)
```

`./test --percentT=false` produces
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; range=0..100 by -128 align 0
by_=        -128:int(8)    ; align_=         127:int(8)    ; range=0..100 by -128 align 127
by_=        -128:int(8)    ; align_=         255:uint(8)   ; range=0..100 by -128test.chpl:22: error: halt reached - casting int(8) less than 0 to uint(8)
```

#### Printing `_alignment`

In trying to debug this issue with @vasslitvinov, we tried printing out `r._alignment` in the format string as well. Replacing the loop body with:

```chapel
if percentT then 
  writef("by_=%12n:%-10s; align_=%12n:%-10s; r._alignment=%12n; range=%t\n",  by_, by_.type:string, align_, align_.type:string, r._alignment, r);
else { 
  writef("by_=%12n:%-10s; align_=%12n:%-10s; r._alignment=%12n; range=",  by_, by_.type:string, align_, align_.type:string, r._alignment);
  writeln(r);
}
```

Now the output looks like:
`./test` produces 
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; r._alignment=           0; range=0..100 by 0 align $CHPL_HOME/modules/internal/ChapelRange.chpl:3079: error: halt reached - Attempt to compute a modulus by zero
```

`./test --percentT=false` produces
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; r._alignment=           0; range=0..100 by -128 align 0
by_=        -128:int(8)    ; align_=         127:int(8)    ; r._alignment=         127; range=0..100 by -128 align 127
by_=        -128:int(8)    ; align_=         255:uint(8)   ; r._alignment=         127; range=0..100 by -128 align 127
by_=         127:int(8)    ; align_=        -128:int(8)    ; r._alignment=         126; range=0..100 by 127 align 126
by_=         127:int(8)    ; align_=         127:int(8)    ; r._alignment=           0; range=0..100 by 127
by_=         127:int(8)    ; align_=         255:uint(8)   ; r._alignment=           1; range=0..100 by 127 align 1
by_=         127:uint(8)   ; align_=        -128:int(8)    ; r._alignment=         126; range=0..100 by 127 align 126
by_=         127:uint(8)   ; align_=         127:int(8)    ; r._alignment=           0; range=0..100 by 127
by_=         127:uint(8)   ; align_=         255:uint(8)   ; r._alignment=           1; range=0..100 by 127 align 1
```

#### Printing `_alignment` in Chapel 1.30

Just to be consistent, running the above example in 1.30 as well
`./test` produces 
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; r._alignment=        -128; range=0..100 by -128 align 0
by_=        -128:int(8)    ; align_=         127:int(8)    ; r._alignment=         127; range=0..100 by -128 align 127
by_=        -128:int(8)    ; align_=         255:uint(8)   ; r._alignment=         255; range=0..100 by -128test.chpl:22: error: halt reached - casting int(8) less than 0 to uint(8)
```

`./test --percentT=false` produces
```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; r._alignment=        -128; range=0..100
by_=        -128:int(8)    ; align_=         127:int(8)    ; r._alignment=         127; range=0..100
by_=        -128:int(8)    ; align_=         255:uint(8)   ; r._alignment=         255; range=0..100
by_=         127:int(8)    ; align_=        -128:int(8)    ; r._alignment=        -128; range=0..100
by_=         127:int(8)    ; align_=         127:int(8)    ; r._alignment=         127; range=0..100
by_=         127:int(8)    ; align_=         255:uint(8)   ; r._alignment=         255; range=0..100
by_=         127:uint(8)   ; align_=        -128:int(8)    ; r._alignment=        -128; range=0..100
by_=         127:uint(8)   ; align_=         127:int(8)    ; r._alignment=         127; range=0..100
by_=         127:uint(8)   ; align_=         255:uint(8)   ; r._alignment=         255; range=0..100
```

### Configuration Information

- Output of `chpl --version`:
```
chpl version 1.31.0 pre-release (18be08a1a5)
  built with LLVM version 14.0.6
  available LLVM targets: amdgcn, r600, nvptx64, nvptx, aarch64_32, aarch64_be, aarch64, arm64_32, arm64, x86-64, x86
Copyright 2020-2023 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: arm64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: fifo
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: cstdlib
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: none
CHPL_RE2: bundled
CHPL_LLVM: bundled *
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: arm64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Tagging @benharsh in this

**Comment by mppf** [comment]
Does it do the same thing when using the C backend (e.g. `CHPL_TARGET_COMPILER=gnu` or `CHPL_TARGET_COMPILER=clang`) ?

**Comment by jabraham17** [comment]
It does not have the same problem using the C backed. It always prints the second version.

```
Testing signed int(8) and unsigned uint(8)
by_=        -128:int(8)    ; align_=        -128:int(8)    ; range=0..100 by -128 align 0
by_=        -128:int(8)    ; align_=         127:int(8)    ; range=0..100 by -128 align 127
by_=        -128:int(8)    ; align_=         255:uint(8)   ; range=0..100 by -128 align 127
by_=         127:int(8)    ; align_=        -128:int(8)    ; range=0..100 by 127 align 126
by_=         127:int(8)    ; align_=         127:int(8)    ; range=0..100 by 127
by_=         127:int(8)    ; align_=         255:uint(8)   ; range=0..100 by 127 align 1
by_=         127:uint(8)   ; align_=        -128:int(8)    ; range=0..100 by 127 align 126
by_=         127:uint(8)   ; align_=         127:int(8)    ; range=0..100 by 127
by_=         127:uint(8)   ; align_=         255:uint(8)   ; range=0..100 by 127 align 1
```

Based on our nightly tests, I also tried with `--fast` and `--no-checks` (with and without the backend). The C backend always prints the correct result, while the LLVM backend with `--fast` or `--no-checks` prints different results inconsistently.

**Comment by mppf** [comment]
It could be an issue with how we are generating the LLVM IR. LLVM IR does not distinguish between signed and unsigned in the types (but rather in the operations applied to them), and it wouldn't surprise me if there is a bug related to our tracking of signed and unsigned.

I think the next step would be to figure out which function is behaving differently in the C and LLVM backends. That can be done by simultaneously running `gdb` on two versions of the program and inspecting the elements that should be the same (arguments to function calls etc). Once we know which function is behaving differently, we can look closely at the LLVM IR we are generating & might see something wrong.

**Comment by jabraham17** [comment]
As a checkpoint for this, here is where I am at.

I reduced the test case to the following which shows different behavior for LLVM and C

```chapel
proc causeProblems(align_, by_, r) {
  writef("by_=%12n:%-10s; align_=%12n:%-10s; range=%t\n",  by_, by_.type:string, align_, align_.type:string, r);
}
var by_ = -128:int(8);
var align_=127:int(8);
var r = (0:align_.type)..(100:align_.type) by by_ align align_;
causeProblems(align_, by_, r);
```

Running the LLVM version through gdb didn't get me very far as it was hard to correlate assembly to Chapel (or even to IR). I did place a few `extern printf`s around the code base and get a little further.
```
extern proc printf(fmt: c_string, x:c_int, y:c_int): c_int;
printf(c"(%d %d)\n",  r._stride:c_int, r._alignment:c_int);
```

In `fileWriter.writef`, `r._alignment` is fine but once it is passed to `_writefOne` it gets corrupted. I looked at the optimized LLVM IR at this point and found a `i32 poison` value.

**Comment by jabraham17** [comment]
I have an even smaller reproducer that does not include writef

```chapel
var by_ = -128:int(8);
var align_=127:int(8);
var r = (0:align_.type)..(100:align_.type) by by_ align align_;
assert(r._alignment == 127);
```

The assert fails when using the LLVM backend compiled with `--fast`. I still see the poison value being added in the optimized version. My current thinking is that LLVM is deciding there is an illegal operation and "optimizing" by inserting poison

**Comment by jabraham17** [comment]
I found the issue! And a solution for this particular problem, but I would be curious if others think a more general solution is required.

TL;DR: LLVM generates the `srem` instruction for the `chpl__mod` function, which has the possibility for undefined behavior with overflow. This problem did not occur with C because C performs a modulo on `int8` with `int16` precision. This can be solved by increasing the bit width for the computation `chpl__mod`.


Currently, range._alignment is the result of `chpl__mod(stride, alignment)`. This modulo operation is done at the given bit width of the operands, in this particular case `int(8)`.  The LLVM IR generated accomplishes this with a `srem` on i8's. Investigating this instruction (https://llvm.org/docs/LangRef.html#srem-instruction), it has the possibility for undefined behavior at integral min/max where overflow may occur. Notably:
> Overflow also leads to undefined behavior; this is a rare case, but can occur, for example, by taking the remainder of a 32-bit division of -2147483648 by -1. (The remainder doesn’t actually overflow, but this rule lets srem be implemented using instructions that return both the result of the division and the remainder.)


I wanted to see what happens if I write this in C. Compiling the following in C to LLVM IR:
```c
char mod(char a, char s) {
  return a%s;
}
long long mod2(long long a, long long s) {
  return a%s;
}
```

The IR for `mod2` is a single instruction, `srem` on `i64`. But the IR for `mod` sign extends the `i8`'s to `i6`, then performs the `srem` on `i16`s, and then truncates the result back to `i8`. (For those who like to read LLVM IR)
```llvm
define dso_local signext i8 @mod(i8 noundef signext %0, i8 noundef signext %1) local_unnamed_addr #0 {
  %3 = sext i8 %0 to i16
  %4 = sext i8 %1 to i16
  %5 = srem i16 %3, %4
  %6 = trunc i16 %5 to i8
  ret i8 %6
}
```

This circumvents the problem of overflow but just performing the operation at a higher bit width. The `i64` case seems to be a special case in the rules for overflow here, because the generated code works at the boundaries.

So the solution for this in the `chpl__mod` specific case is to just cast to a bigger type. Example code:
```chapel
proc chpl__mod(dividend:integral, modulus:integral)
{
    const dividend_ = if isUintValue(dividend) then dividend:uint else dividend:int;
    const modulus_ = if isUintValue(modulus) then modulus:uint else modulus:int;
    const m = abs(modulus_);

    var tmp = dividend_ % m;
    if isInt(dividend_) then
      if tmp < 0 then tmp += m;

    return tmp:dividend.type;
}
```
> Note: This code is a first draft and contains more checks than needed. For instance, I believe this issue only occurs with signed numbers (at the very least I see no comments in the docs about uint and my brief experiments with C show that it does not do the same widen for unsigned types). So we don't really need the upcast to `uint`, which may save a few instructions in the assembly. 

However, this would be an issue elsewhere if a user uses `%` on `int(8)`, `int(16)`, or `int(32)`, since by default the compiler will not add in the sign extensions/truncations.

**Comment by mppf** [comment]
Thanks for tracking it down @jabraham17 !

Another possibility is to adjust the LLVM backend to do i8 mods with i16 width (like C seems to).

I think it's important to point out that `chpl__mod` is primarily used in `ChapelRange`  (also in some distributions). So we can make it to what makes sense for ranges. I am pretty sure we don't want these `int(8)` ranges to create undefined behavior, but I'm also not certain what we want to do in this case. Historically, we have made some exceptions for ranges with minimal/maximal bounds.

Is it accurate that in the failing case we are computing `chpl__mod(-128:int(8), 127:int(8))` ?

I suspect it'd be possible to write `chpl__mod` in terms of `uint(t)` values combined with sign information.

Also, while I think it is OK for us to match C in having `%` have some undefined behavior, I suspect that `proc mod` in AutoMath should handle `mod(-128:int(8), 127:int(8))`, so perhaps it needs some kind of adjustment as well. If `proc mod` handled it, `chpl_mod` could call it.

**Comment by jabraham17** [comment]
This bug ended up being an issue with `abs(Types.min)`, which LLVM would interpret as invalid and insert a `poison` value. Resolved with #22274

**Comment by bradcray** [comment]
Nice sleuthing!!
